### ISSUE TITLE
Gitea: integration with Draw.io

### ISSUE THREAD

**Issue by AKopytenko** [go-gitea/gitea#28903]
### Feature Description

Hi there!

I have an idea for integrating Gitea services and Draw.io. In my opinion, this can be a pretty useful thing!

"The user attaches the .drawio file to the messages in the issues. When clicking on the attached file, the user is redirected to the draw.io, at the same time, the contents of the file [XML in base64 or the like] are sent to the payload of the POST request. Draw.io receiving the file content and expands it into a schema. Apparently, the issue is complex and requires collaboration with the guys from Gitea. Let's try?" (с)

https://github.com/jgraph/drawio/issues/4128

**Comment by techknowlogick** [comment]
I wonder if this could be achieved with an external renderer via the config so nothing new needs to be added to Gitea

**Comment by lunny** [comment]
> I wonder if this could be achieved with an external renderer via the config so nothing new needs to be added to Gitea

I don't think it's possible to be as an external renderer. Because it needs to read/save the files. It should be more like the markdown editor of course for .drawio files but .md files.

**Comment by lunny** [comment]
According to the statement on their README.
> draw.io is not suitable as a framework for building other products from. For this try either [Tldraw](https://github.com/tldraw/tldraw) or [Excalidraw](https://github.com/excalidraw/excalidraw).

Tldraw is non-commercial license. Looks like https://github.com/excalidraw/excalidraw is the only possible solution.

**Comment by silverwind** [comment]
As I understand it, the only official interface is the full-blown editor UI. What we need would be a viewer-only UI. Maybe there are third-party made renderes available somewhere.

**Comment by silverwind** [comment]
Seems confluence has done some integration with a `viewer.min.js` file from the draw.io repo, so it may be possible:

https://www.drawio.com/doc/faq/configure-javascript-viewer-drawio-confluence-server

**Comment by tobhv** [comment]
gitlab supports viewing, editing draw.io files. seems to work offline from draw.io some links:

- some code in master: https://gitlab.com/gitlab-org/gitlab/-/tree/master/app/assets/javascripts/drawio?ref_type=heads
- and merge request that looks like mvp: https://gitlab.com/gitlab-org/gitlab/-/merge_requests/86206/diffs#7981eac67e1980a41d95e5bfc2ccea7350ce53ae 

i think editing is a very nice feature.

**Comment by silverwind** [comment]
Seems they are embedding a iframe from `https://embed.diagrams.net`, not an ideal solution but could at least work with self-hosted renderer if that URL is made configurable.

**Comment by lafriks** [comment]
Why not just use drawio JS as lib?

**Comment by silverwind** [comment]
> Why not just use drawio JS as lib?

There is no lib as far as I know, it's only available integrated with their backend.

**Comment by mbnoimi** [comment]
+1
I missed this feature after migrating my repositories from Gltlab into Gitea

**Comment by lunny** [comment]
This could be a plugin after #34917 is implemented

**Comment by silverwind** [comment]
Yes, but it would still require a external render server. The JS could render a `<iframe>`. Thought I guess it may be easier to just support this iframe-rendering first-party, it's probably only a few lines of code.

**Comment by silverwind** [comment]
According to https://github.com/jgraph/drawio/discussions/3430#discussioncomment-5322399, it may be possible to render `.drawio` to HTML with the viewer JS.

**Comment by Rob4github** [comment]
I wrote a custom renderer based on this Script:

based on this thread: https://github.com/jgraph/drawio/discussions/3430#discussioncomment-5322399

To reproduce add this block in your files under /conf/app.ini

```ini
[markup.drawio]
ENABLED = true
FILE_EXTENSIONS = .drawio,.dio
RENDER_COMMAND = /data/gitea/scripts/render-drawio.sh
IS_INPUT_FILE = false
RENDER_CONTENT_MODE = iframe
```

And a new file named "render-drawio.sh" under /scripts/

```bash
#!/bin/bash

# Gitea external renderer for Draw.io files

XML_DATA=$(cat | sed 's/\\/\\\\/g' | sed 's/`/\\`/g' | sed "s/'/\\\\'/g" | sed 's/\$/\\$/g')

cat << EOF
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <script type="text/javascript" src="https://viewer.diagrams.net/js/viewer-static.min.js"></script>
    <style>
        body { margin: 0; padding: 0; }
        .mxgraph { max-width: 100%; border: 1px solid transparent; }
    </style>
</head>
<body>
    <div id="drawio-diagram"></div>
    <script>
        (function () {
            const chatMap = {
                "&": "&amp;",
                "'": "&#x27;",
                "\`": "&#x60;",
                '"': "&quot;",
                "<": "&lt;",
                ">": "&gt;",
            };
            function replaceMatchedCharacters(match) {
                return chatMap[match];
            }
            function escapeHTML(string) {
                if (typeof string !== "string") return string;
                return string.replace(/[&'\`"<>]/g, replaceMatchedCharacters);
            }
            function createMxGraphData(xml) {
                return {
                    "highlight": "#0000ff",
                    "nav": true,
                    "resize": true,
                    "toolbar": "zoom layers lightbox",
                    "edit": "_blank",
                    "xml": xml
                };
            }
            async function drawioConverterAsync(xml) {
                return new Promise((resolve) => {
                    const mxGraphData = createMxGraphData(xml);
                    const json = JSON.stringify(mxGraphData);
                    const mxGraphHTML = '<div class="mxgraph" style="max-width:100%;border:1px solid transparent;" data-mxgraph="' + escapeHTML(json) + '"></div>';
                    resolve(mxGraphHTML);
                });
            }
            const xmlData = \`${XML_DATA}\`;
            drawioConverterAsync(xmlData).then((content) => {
                const graphContainer = document.getElementById('drawio-diagram');
                graphContainer.innerHTML = content;
                window.GraphViewer.processElements();
            }).catch((err) => {
                console.error("Error loading draw.io file:", err);
                document.getElementById('drawio-diagram').innerHTML = '<p style="color:red;">Error loading diagram: ' + err.message + '</p>';
            });
        })();
    </script>
</body>
</html>
EOF
```
